File: COLLEGE/ASSIGNMENT7/c/Q.7.c
#include <stdio.h>
#include <stdlib.h>

struct Queue {
    int* arr;
    int front, rear, size, capacity;
};

struct Queue* createQueue(int capacity) {
    struct Queue* queue = (struct Queue*)malloc(sizeof(struct Queue));
    queue->capacity = capacity;
    queue->front = queue->size = 0;
    queue->rear = capacity - 1;
    queue->arr = (int*)malloc(queue->capacity * sizeof(int));
    return queue;
}

int isFull(struct Queue* queue) {
    return (queue->size == queue->capacity);
}

int isEmpty(struct Queue* queue) {
    return (queue->size == 0);
}

void enqueue(struct Queue* queue, int item) {
    if (isFull(queue)) return;
    queue->rear = (queue->rear + 1) % queue->capacity;
    queue->arr[queue->rear] = item;
    queue->size++;
}

int dequeue(struct Queue* queue) {
    if (isEmpty(queue)) return -1;
    int item = queue->arr[queue->front];
    queue->front = (queue->front + 1) % queue->capacity;
    queue->size--;
    return item;
}

struct Stack {
    struct Queue* q1;
    struct Queue* q2;
};

struct Stack* createStack(int capacity) {
    struct Stack* stack = (struct Stack*)malloc(sizeof(struct Stack));
    stack->q1 = createQueue(capacity);
    stack->q2 = createQueue(capacity);
    return stack;
}

void push(struct Stack* stack, int item) {
    enqueue(stack->q2, item);
    while (!isEmpty(stack->q1)) {
        enqueue(stack->q2, dequeue(stack->q1));
    }
    struct Queue* temp = stack->q1;
    stack->q1 = stack->q2;
    stack->q2 = temp;
}

int pop(struct Stack* stack) {
    return dequeue(stack->q1);
}

void display(struct Stack* stack) {
    if (isEmpty(stack->q1)) {
        printf("Stack is empty\n");
        return;
    }
    printf("Stack elements: ");
    for (int i = stack->q1->front; i < stack->q1->size + stack->q1->front; i++) {
        printf("%d ", stack->q1->arr[i % stack->q1->capacity]);
    }
    printf("\n");
}

int main() {
    struct Stack* stack = createStack(5);
    push(stack, 10);
    push(stack, 20);
    push(stack, 30);
    printf("Popped: %d\n", pop(stack));
    display(stack);
    return 0;
}

File: COLLEGE/ASSIGNMENT7/python/Q.7.py
from collections import deque

class StackUsingQueues:
    def __init__(self):
        self.queue1 = deque()  # Main queue to store elements
        self.queue2 = deque()  # Auxiliary queue for stack operations

    # Function to push an element onto the stack
    def push(self, item):
        # Add the new item to the auxiliary queue
        self.queue2.append(item)
        
        # Move all elements from queue1 to queue2
        while self.queue1:
            self.queue2.append(self.queue1.popleft())
        
        # Swap the names of the two queues
        self.queue1, self.queue2 = self.queue2, self.queue1
        print(f"Pushed {item} onto the stack.")

    # Function to pop the top element from the stack
    def pop(self):
        if self.is_empty():
            print("Stack is empty! Cannot pop element.")
            return None
        popped_item = self.queue1.popleft()  # Remove the front element from queue1
        print(f"Popped {popped_item} from the stack.")
        return popped_item

    # Function to peek at the top element of the stack
    def peek(self):
        if self.is_empty():
            print("Stack is empty! Cannot peek.")
            return None
        return self.queue1[0]  # Return the front element without removing it

    # Function to check if the stack is empty
    def is_empty(self):
        return len(self.queue1) == 0

    # Function to display the current elements in the stack
    def display(self):
        if self.is_empty():
            print("Stack is empty!")
        else:
            print("Stack elements:", list(self.queue1))  # Convert deque to list for display


# Test the StackUsingQueues class
stack = StackUsingQueues()

# Push elements onto the stack
stack.push(10)
stack.push(20)
stack.push(30)

# Display the stack
stack.display()

# Peek at the top element
print(f"Top element: {stack.peek()}")

# Pop elements from the stack
stack.pop()
stack.pop()

# Display the stack after popping
stack.display()

# Push more elements
stack.push(40)
stack.push(50)

# Final state of the stack
stack.display()
